DebuggingFramework.md

```diff
@@ -0,0 +1,139 @@
+#WebAssembly Debugging
+
+This is a design proposal for enabling source-level debugging of WebAssembly
+modules.
+
+## Introduction
+
+There are several entities involved in debugging a source program compiled into
+WebAssembly:
+
+- the compiled code being executed, called the _debuggee_
+- the browser engine executing the debuggee
+- the browser devtools
+- the source-program listing, usually split into multiple source files
+- the description of how the debuggee code maps to the source program and vice
+  versa (e.g., WebAssembly code locations to source code locations, source
+  variables to WebAssembly memory, WebAssembly call stack to source functions,
+  etc.), called the _debug info_
+
+These entities carry information and interact in particular ways to enable a
+user to perform debugging actions at the source-code level.  For example, the
+browser devtools offer a UI allowing the user to step through the source code,
+requiring partial execution of the debuggee chunk corresponding to each step,
+and displaying the source file containing the stepped-through program.
+
+Specifying the debugging framework means describing how these entities interact,
+what they need from each other, and how they access it; all in order to execute
+the user's commands.
+
+## Debugger as a Procedural Interface
+
+A crucial element of this design proposal is to omit specifying a particular
+format for the debug info.  Instead, we introduce a separate entity we call the
+debugger; its job is to act as a procedural intermediary between the devtools
+and the debug info.  Devtools execute user actions by leveraging debugger
+methods we will specify here.
+
+Thus any debug-info format is acceptable, as long as the debugger can correctly
+interpret it.  The format is, in effect, a convention between the front-end
+parsing the source and the debugger.
+
+What the design _does_ specify is two interaction protocols:
+
+1. between the devtools and the debugger, and
+2. between the debugger and the wasm execution engine
+
+## Action Flow
+
+For reference, the entities are charted in the figure below:
+
+![Entities making up the debug framework](wasm-debug-chart.png)
```

Comment:
Actually, this aims to specify 3 and 5, while leaving 4 unspecified (other than saying how the debugger can find the debug info and source files).  The way the debugger interprets the debug info will be hidden from other actors in this picture, freeing the debugger implementers to choose whatever suits them.